pkg/server/server.go

```diff
@@ -988,3 +992,127 @@ func tokenToObjectName(token string) string {
 	h := sha256.Sum256([]byte(name))
 	return sha256Prefix + base64.RawURLEncoding.EncodeToString(h[0:])
 }
+
+func (s *Server) UpdateServiceAccountCertAndTokenPeriodically(oidcClientConfig *auth.Config, caCertFilePath, k8sInClusterBearerToken string) {
+	for {
+		time.Sleep(10 * time.Minute)
+		klog.Info("Updating service account token certificate and token...")
+
+		token, err := GetInClusterToken(k8sInClusterBearerToken)
+		if err != nil {
+			klog.Fatalf("Error updating service account token: %v", err)
+		}
+		tlsConfig, err := GetInClusterTLSConfig(caCertFilePath)
+		if err != nil {
+			klog.Fatalf("Error updating service account certificate: %v", err)
+		}
+
+		s.ServiceAccountToken = token
+		s.K8sProxyConfig.TLSClientConfig = tlsConfig
+		s.K8sClient = &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: tlsConfig,
+			},
+		}
+
+		switch s.K8sAuthType {
+		case "service-account":
+			s.StaticUser = &auth.User{
+				Token: token,
+			}
+		case "oidc", "openshift":
+			s.ServiceAccountToken = token
+		}
+		klog.Info("Updating listers...")
+		s.SetListers()
+
+		klog.Info("Updating authenticator...")
+		s.UpdateClientConfigAuther(oidcClientConfig)
+	}
+}
+
+func (s *Server) SetListers() {
+	s.MonitoringDashboardConfigMapLister = s.GetResourceLister(
+		"/api/v1/namespaces/openshift-config-managed/configmaps",
+		"console.openshift.io/dashboard=true",
+		nil,
+	)
+
+	s.KnativeEventSourceCRDLister = s.GetResourceLister(
+		"/apis/apiextensions.k8s.io/v1/customresourcedefinitions",
+		"duck.knative.dev/source=true",
+		knative.EventSourceFilter,
+	)
+
+	s.KnativeChannelCRDLister = s.GetResourceLister(
+		"/apis/apiextensions.k8s.io/v1/customresourcedefinitions",
+		"duck.knative.dev/addressable=true,messaging.knative.dev/subscribable=true",
+		knative.ChannelFilter,
+	)
+}
+
+func (s *Server) GetResourceLister(
+	path,
+	labelSelector string,
+	respFilter FilterFunction,
+) ResourceLister {
+	httpClientTransport := &http.Transport{
+		TLSClientConfig: s.K8sProxyConfig.TLSClientConfig,
+	}
+	// Setting the proxy for all the listeners
+	if s.K8sMode == "off-cluster" {
+		httpClientTransport.Proxy = http.ProxyFromEnvironment
+	}
+
+	return NewResourceLister(
+		s.ServiceAccountToken,
+		&url.URL{
+			Scheme: s.K8sProxyConfig.Endpoint.Scheme,
+			Host:   s.K8sProxyConfig.Endpoint.Host,
+			Path:   s.K8sProxyConfig.Endpoint.Path + path,
+			RawQuery: url.Values{
+				"labelSelector": {labelSelector},
+			}.Encode(),
+		},
+		&http.Client{
+			Transport: httpClientTransport,
+		},
+		respFilter,
+	)
+}
+
+func (s *Server) UpdateClientConfigAuther(oidcClientConfig *auth.Config) {
+	var err error
+	if s.Authers[serverutils.LocalClusterName], err = auth.NewAuthenticator(context.Background(), oidcClientConfig); err != nil {
+		klog.Fatalf("Error updating authenticator: %v", err)
+	}
+}
+
+func GetInClusterToken(tokenPath string) (string, error) {
+	token, err := os.ReadFile(tokenPath)
```

Comment:
consider using `transport.CachedFileTokenSource`

https://github.com/kubernetes/kubernetes/blob/7f532891c9c42500f75939fd931596cd42b1765b/staging/src/k8s.io/client-go/transport/token_source.go#L69-L70

Reply:
Used the `NewBearerAuthWithRefreshRoundTripper()` method in the `GetK8sClient()` function instead.